Initial PocketBeagle 2 MSPM0L1105 support

Ayush1325 opened the pull request:
PocketBeagle 2 contains an MSPM0L1105 soc which emulates ADC and EEPROM.
It is connected to the AM62xx soc over I2C and SPI.

The initial support just allows building the minimal zephyr sample. Due
to the constraints and the purpose of the SOC, most peripherals should
not be enabled in the main dts.

Ayush1325 committed:
soc: ti: mspm0: mspm0l: Selectively enable MPU

- Not all MSPM0L series socs seem to have MPU. After a glance at the TI
  website for features, only MSPM0L222x class socs seem to have MPU.

Signed-off-by: Ayush Singh <ayush@beagleboard.org>

Ayush1325 committed:
drivers: clock_control: mspm0: Check SOC for HFCLK

- The function DL_SYSCTL_setHFCLKSourceHFCLKIN is not defined for
  MSPM0L11XX and MSPM0L13XX.

Signed-off-by: Ayush Singh <ayush@beagleboard.org>

Ayush1325 commented:
cc @parthitce

parthitce reviewed (changes requested):
Thanks for your efforts!

Ayush1325 commented on drivers/clock_control/clock_control_mspm0.c:
```diff
@@ -37,6 +37,11 @@
 #define MSPM0_PLL_ENABLED 1
 #endif
 
+#if !(CONFIG_SOC_MSPM0L1105 || CONFIG_SOC_MSPM0L1106 || CONFIG_SOC_MSPM0L1303 ||                   \
```
You mean something like `DT_NODE_EXISTS(hfxt)`?

parthitce replied:
yeah, something like that. This way hardware presence is always part of dts.

Ayush1325 commented:
Not really sure where I should select `CPU_HAS_ARM_MPU`. Maybe I can define something like SOC_MSPM0L222X in the `Kconfig` and do the mpu selection on it?

Ayush1325 commented:
> If not the whole series, then then only placeholder IMO is respective SoC in Kconfig.soc.

You mean how I have done it right now? That seems to trigger a compliance error.

ssekar15 commented:
> > > If not the whole series, then then only placeholder IMO is respective SoC in Kconfig.soc.
> > 
> > 
> > You mean how I have done it right now? That seems to trigger a compliance error.
> 
> Yes, but not sure why the compliance is not happy about it. May be @ssekar15 ?

Compliance check treats Kconfig and Kconfig.soc separately and expects `CPU_HAS_ARM_MPU` has to be defined in Kconfig.soc.

ssekar15 commented:
> > Compliance check treats Kconfig and Kconfig.soc separately and expects `CPU_HAS_ARM_MPU` has to be defined in Kconfig.soc.
> 
> So, any idea how to make compliance check happy?

@carlescufi could you share your thought that compliance check has to be improved? Kconfig definition on ARCH is not taken account for SoC level Kconfig

jadonk reviewed (changes requested):
The organization of the series having an MPU or not seems to need to be addressed here.

Also, I think some documentation on the GPIOs that could be utilized for SWD is justified. Further, it isn't clear to me how this SoC enables SWD.

Finally, enabling `west flash` is something that could be contributed separately, but please do make a note to do it. Documenting the Linux firmware update process in code makes more sense to me than the lines documented here. We made a bit of a skeleton for this in our effort to enable `west flash` for the R5s on BeagleY-AI over remoteproc and I think it should be a high priority to enable that for each target CPU (outside the A53s) to enable clean native (A53-hosted) development.

Ayush1325 commented on boards/beagle/pocketbeagle_2/doc/index.rst:
```diff
@@ -6,8 +6,6 @@ Overview
 PocketBeagle 2 is a computational platform powered by TI AM62x SoC (there are two
 revisions, AM6232 and AM6254).
```
Added.

Ayush1325 commented on boards/beagle/pocketbeagle_2/doc/index.rst:
```diff
@@ -17,7 +15,9 @@ cluster with an Arm Cortex-M4F microcontroller, Imagination Technologies AXE-1-1
-Zephyr is ported to run on the both A53 cores and/or M4F core.
+Additionally, PocketBeagle 2 also contains an MSPM0L1105 SoC which serves as EEPROM and ADC.
+
+Zephyr is ported to run on the both A53 cores and/or M4F core on AM62x and MSPM0L1105.
```
Added

carlescufi commented:
> > > Compliance check treats Kconfig and Kconfig.soc separately and expects `CPU_HAS_ARM_MPU` has to be defined in Kconfig.soc.
> > 
> > 
> > So, any idea how to make compliance check happy?
> 
> @carlescufi could you share your thought that compliance check has to be improved? Kconfig definition on ARCH is not taken account for SoC level Kconfig

@nordicjm could you please take a look?

nordicjm commented on soc/ti/mspm0/mspm0l/Kconfig.soc:
```diff
@@ -63,10 +63,12 @@ config SOC_MSPM0L1346
 config SOC_MSPM0L2227
 	bool
 	select SOC_SERIES_MSPM0L
+	select CPU_HAS_ARM_MPU
```
this has no place here

Ayush1325 replied:
So, where and how should this be selected? Do we need to spin of MSPM0L222x as a separate series?

nordicjm replied:
put it in `Kconfig`. As per https://docs.zephyrproject.org/latest/hardware/porting/soc_porting.html#create-your-soc-directory
> Kconfig.soc: the base SoC configuration which defines a Kconfig SoC symbol in the form of config SOC_<soc-name> and provides the SoC name to the Kconfig SOC setting. If the soc.yml describes a SoC family and series, then those must also be defined in this file. Kconfig settings outside of the SoC tree must not be selected. To select general Zephyr Kconfig settings the Kconfig file must be used.

nordicjm commented on soc/ti/mspm0/mspm0l/Kconfig.soc:
```diff
@@ -63,10 +63,12 @@ config SOC_MSPM0L1346
 config SOC_MSPM0L2228
 	bool
 	select SOC_SERIES_MSPM0L
+	select CPU_HAS_ARM_MPU
```
nor here

Ayush1325 committed:
boards: beagle: pocketbeagle_2: Add MSPM0 support

PocketBeagle 2 contains an MSPM0L1105 soc which emulates ADC and EEPROM.
It is connected to the AM62xx soc over I2C and SPI.

The initial support just allows building the minimal zephyr sample. Due
to the constraints and the purpose of the SOC, most peripherals should
not be enabled in the main dts.

Signed-off-by: Ayush Singh <ayush@beagleboard.org>

kartben commented on boards/beagle/pocketbeagle_2/doc/index.rst:
```diff
@@ -3,21 +3,25 @@
-revisions, AM6232 and AM6254).
-
-The board configuration provides support for the ARM Cortex-M4F MCU core.
+PocketBeagle 2 is a computational platform powered by TI AM6254 SoC (rev A0,
+which has been replaced by rev A1 used AM6232).
```
so why are you not defining revision A1 (and I guess make it the default?) in board.yml - did I miss something?

Ayush1325 replied:
Well, I only have rev A0 with me. Theoretically, rev A1 should not require any changes, but well, since I cannot test it myself, I am not sure about adding it.

kartben replied:
Oh my bad, sorry, I thought this PR was effectively introducing support for rev A1 / AM6254, so I clearly read this too quickly :/ That being said, the comment still kind of stands - this wording change (and the update to the section hardware is even worse) doesn't really belong to this commit and is confusing since it's talking about a version that's effectively not supported in Zephyr..?

Ayush1325 replied:
I have rolled back the unrelated changes for now. Will make those in rev A1 PR.

parthitce reviewed (approved):
Thanks for your effort @Ayush1325

jadonk reviewed (approved):
The reorganization of when to define HAS_ARM_MPU looks good to me. Approved.
